### PR TITLE
Clean up type identifiers

### LINKED ISSUE #3583: Clean up type identifiers
Thread from: https://github.com/ethereum/solidity/pull/2728#issuecomment-339281942

### PR DESCRIPTION
**Pull request by axic**
Fixes #3583.

### CODE REVIEW

**Comment by axic** [comment]
@chriseth this is work in progress, need to clean up the commits, but it produces the exact same output as before

File: libsolidity/ast/Types.cpp [hunk 1]
```diff
@@ -165,11 +167,33 @@ string identifierList(TypePointer const& _type1, TypePointer const& _type2)
...
+	return parenthesizeIdentifier(_internal);
+//	return parenthesizeIdentifier(boost::algorithm::replace_all_copy(_internal, "$", "$$$"));
 }
 
 }
 
+string Type::escapeIdentifier(string const& _identifier)
+{
+	string ret = _identifier;
+	boost::algorithm::replace_all(ret, "$", "$$$");
```
Comment (axic) [inline libsolidity/ast/Types.cpp]: Basically this is the current rule and it goes against the description, which says `_$$$_`

**Comment by chriseth** [comment]
Looks good!

**Comment by axic** [comment]
@chriseth added tests but unescaping doesn't work.

How about removing the unescaping part for now and merging the code which is just a refactoring, then separately unescaping and perhaps changing the escape codes so it can be properly unescaped?

**Comment by chriseth** [comment]
Would be good to have unescaping working to see that it actually is a bijection. Let me do the review, then I can tell.

File: libsolidity/ast/Types.cpp [hunk 2]
```diff
@@ -165,11 +165,33 @@ string identifierList(TypePointer const& _type1, TypePointer const& _type2)
...
-	return parenthesizeIdentifier(boost::algorithm::replace_all_copy(_internal, "$", "$$$"));
+	return parenthesizeIdentifier(_internal);
 }
 
 }
 
+string Type::escapeIdentifier(string const& _identifier)
+{
+	string ret = _identifier;
+	// FIXME: should be _$$$_
```
Comment (chriseth) [inline libsolidity/ast/Types.cpp:176]: Why not do that?

**Comment by axic** [comment]
This simple way (since it is just replacements one after the other) of unescaping will break when underscores are next to escaped `$`:

input: `_$_`
escaped: `_$$$_`
unescaped: `_$_` -> `,`

It could be done properly by doing unescaping in one go.

File: test/libsolidity/SolidityTypes.cpp [hunk 3]
```diff
@@ -88,6 +88,36 @@ BOOST_AUTO_TEST_CASE(storage_layout_arrays)
...
+{
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("$_"), "(");
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("_$"), ")");
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("_$_"), ",");
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("$$$"), "$");
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("$__$"), "()");
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("$__$__$"), "(,)");
+	BOOST_CHECK_EQUAL(Type::unescapeIdentifier("$__$_$$$_$__$"), "(,$,)");
+	BOOST_CHECK_EQUAL(
+		Type::unescapeIdentifier("$_$___$__$$$_$$$$$$_$___$_$$$$$$_$__$__$_$$$$$$_$_$_$$$_$__$__$__$_$_$$$$$$$_$_$$$$$$"),
```
Comment (chriseth) [inline test/libsolidity/SolidityTypes.cpp]: Shouldn't the test run escape -> unescape and then compare equal?

**Review by chriseth — APPROVED** [review]

File: libsolidity/ast/Types.cpp (hunk 2 above)
Comment (axic) [reply libsolidity/ast/Types.cpp:176]: I didn't wanted to change the currently identifiers, they were using `$` -> `$$$`

Comment (chriseth) [reply libsolidity/ast/Types.cpp:176]: ah ok!

**Comment by chriseth** [comment]
ok, I'm fine with merging without unescaping.

**Pull request merged**